packages/databricks-vscode/src/bundle/BundleRunManager.ts

```
@@ -0,0 +1,184 @@
+        if (target === undefined) {
+            return false;
+        }
+        const terminalName = this.getTerminalName(target, resourceKey);
+        return this._isRunning.get(terminalName) ?? false;
+    }
+    constructor(
+        private readonly bundleRemoteStateModel: BundleRemoteStateModel
+    ) {}
+
+    getTerminalName(target: string, resourceKey: string) {
+        return `Run ${resourceKey} (${target})`;
+    }
+
+    async run(resourceKey: string) {
+        const target = this.bundleRemoteStateModel.target;
+        if (target === undefined) {
+            throw new Error(`Cannot run ${resourceKey}, Target is undefined`);
+        }
+        const terminalName = this.getTerminalName(target, resourceKey);
+
+        if (!this.terminalDetails.has(terminalName)) {
+            this.terminalDetails.set(
+                terminalName,
+                this.createTerminal(terminalName)
+            );
+        }
+        let terminal = this.terminalDetails.get(terminalName)!;
+
+        const disposables: Disposable[] = [];
+        try {
+            terminal.terminal.show();
+            if (
+                window.terminals.find(
+                    (i) => i.name === terminal?.terminal.name
+                ) === undefined
+            ) {
+                // The terminal has been closed. Recreate everything.
+                terminal = this.createTerminal(terminalName);
+                this.terminalDetails.set(terminalName, terminal);
+            }
+            if (terminal.pty.process !== undefined) {
+                // There is already a process running. Raise error
+                throw new Error(
+                    `Process already running. Pid: ${terminal.pty.process.pid}`
+                );
+            }
+            this.cancellationTokenSources.get(terminalName)?.cancel();
+            this.cancellationTokenSources.get(terminalName)?.dispose();
+            this.cancellationTokenSources.delete(terminalName);
```

Is that even possible to have the cancellation token at this point in time? Would that mean that a process is already running and we should throw an error instead, as we do several lines above?